packages/codemirror/src/extension.ts

```diff
@@ -0,0 +1,877 @@
+     */
+    constructor(builder: (value: T) => Extension) {
+      this._compartment = new Compartment();
+      this._builder = builder;
+    }
+
+    /**
+     * Create an editor extension for the provided value.
+     *
+     * @param value Editor extension parameter value
+     * @returns The editor extension
+     */
+    instance(value: T): Extension {
+      return this._compartment.of(this._builder(value));
+    }
+
+    /**
+     * Reconfigure an editor extension.
+     *
+     * @param value Editor extension value
+     * @returns Editor state effect
+     */
+    reconfigure(value: T): StateEffect<T> {
+      return this._compartment.reconfigure(
+        this._builder(value)
+      ) as StateEffect<T>;
+    }
+
+    private _compartment: Compartment;
+    private _builder: (value: T) => Extension;
+  }
+
+  /**
+   * Immutable editor extension class
+   */
+  class ImmutableExtension implements IConfigurableExtension<undefined> {
+    /**
+     * Create an immutable editor extension.
+     *
+     * @param extension Extension
+     */
+    constructor(extension: Extension) {
+      this._extension = extension;
+    }
+
+    /**
+     * Create an editor extension.
+     *
+     * @returns The editor extension
+     */
+    instance(): Extension {
+      return this._extension;
+    }
+
+    /**
+     * Reconfigure an editor extension.
+     *
+     * This is a no-op
+     */
+    reconfigure(): null {
+      // This is a no-op
+      return null;
+    }
+
+    private _extension: Extension;
+  }
+
+  /**
+   * Creates a dynamically configurable editor extension.
+   *
+   * @param builder Extension builder
+   * @return The extension
+   */
+  export function createConfigurableExtension<T>(
+    builder: (value: T) => Extension
+  ): IConfigurableExtension<T> {
+    return new ConfigurableExtension<T>(builder);
+  }
+
+  /**
+   * Creates a configurable extension returning
+   * one of two extensions depending on a boolean value.
+   *
+   * @param truthy Extension to apply when the parameter is true
+   * @param falsy Extension to apply when the parameter is false
+   * @return The extension
+   */
+  export function createConditionalExtension(
+    truthy: Extension,
+    falsy: Extension = []
+  ): IConfigurableExtension<boolean> {
+    return new ConfigurableExtension<boolean>(value =>
+      value ? truthy : falsy
+    );
+  }
+
+  /**
+   * Creates an immutable extension.
+   *
+   * @param extension Immutable extension
+   * @return The extension
+   */
+  export function createImmutableExtension(
+    extension: Extension
+  ): IConfigurableExtension<undefined> {
+    return new ImmutableExtension(extension);
+  }
+
+  /**
+   * Get the default editor extensions.
+   *
+   * @returns CodeMirror 6 extension factories
+   */
+  export function getDefaultExtensions(
+    options: {
+      themes?: IEditorThemeRegistry;
+      translator?: ITranslator | null;
+    } = {}
+  ): IEditorExtensionFactory<any>[] {
```

Comment:
```suggestion
  ): ReadonlyArray<Readonly<IEditorExtensionFactory<any>>> {
```

I'm not sure what the answer to this question is: If you switch to `ReadonlyArray` of `Readonly` objects, are the `Object.freeze` calls still necessary?

Reply:
As far as the [documentation](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) goes, it does not modify the generated code. So we should keep the `Object.freeze` (the documentation example snippet is actually about that). But definitely your suggestion is better reflecting the implementation.